File: Companies/VMware/LongestStrChain.java
package Companies.VMware;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class LongestStrChain {
    public int longestStrChain(String[] words) {
        Map<String, Integer> m = new HashMap<>();
        int re = 0;
        Arrays.sort(words, (a,b)->a.length()-b.length());
        for (String word : words) {
            int len = 0;
            for (int i = 0; i < word.length(); i++) {
                String temp = word.substring(0, i) + word.substring(i+1);
                len = Math.max(len, m.getOrDefault(temp, 0)+1);
            }
            re = Math.max(re, len);
            m.put(word, len);
        }
        return re;
    }
}

File: Companies/VMware/LongestStrChain.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class LongestStrChain
{
public:
  int longestStrChain(vector<string> &words)
  {
    sort(words.begin(), words.end(), [](string x, string y) { return x.length() < y.length(); });
    unordered_map<string, int> m{};
    int re{0};
    for (string word : words)
    {
      for (int i{0}; i < word.length(); i++)
      {
        m[word] = max(m[word], m[word.substr(0, i) + word.substr(i + 1)] + 1);
      }
      re = max(re, m[word]);
    }
    return re;
  }
};